url_shortener_proj/app/Http/Controllers/UrlController.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Url;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\DB;
+use Str;
+use App\Visitor;
+use Hash;
+use Session;
+
+
+class UrlController extends Controller
+{
+    public function store(Request $request)
+    {
+        $request->validate([
+            'original' => 'required|url'
+        ]);
+        $url = Url::create([
+            'original' => $request->original,
+            'short' => Str::random(7)
+        ]);
+        $short = $url->short;
+
+        Session::flash('success', 'The short URL was generated successfully!');
+
+        return redirect()->route('urls.details', $short);
+    }
+
+    public function details($short)
```

Comment:
If generated short string is `ABCD123`, can you please tell me:

1) What will be the short URL?
2) What will be the details URL?

Reply:
1. Short URL will be **ABCD123,**
2. Details URL will be **localhost:8000/urls/ABCD123**

Reply:
So, if I type `ABCD123` in my browser's address bar, I will be redirected to the original url?

Reply:
I honestly think that's the answer on your questions, atleast the second one. If I am wrong, can you give me some more feedbacks about what should be sent to friends as a short URL to paste in their address bar? Thank You!

Reply:
Let's start with the really long URL we want to share - http://www.thelongestdomainnameintheworldandthensomeandthensomemoreandmore.com/. This domain itself has 79 characters. If we want to link to specific page on this site, the URL will be much longer, way over 100 characters. That's very long, especially if you want to share it on Twitter, where tweets have the limit of 280 characters.

So, we want to create shorter alternative instead - that's why we need to use URL shorteners. The goal of these shorteners are simple - given the original, long URL, output much shorter URL which will, once visited, redirect user back to the original address. As an added bonus, these shorteners can usually track number of visits, so that original author how many people clicked on his short URL.

We are building one of these URL shortener services. Once we complete it, we can host it on the internet and assign some domain name - for example: http://sho.rt. This means that, once you visit http://sho.rt, you'll get that homepage where you can input your URL and (optionally) email address. So far, so good.

You now decide to shorten that long domain name - http://www.thelongestdomainnameintheworldandthensomeandthensomemoreandmore.com/ - and you paste it into URL input. Once you click on "Generate" button, app will create new row in the database and write `ABCD123` in `short` column. After that, you will be redirected to details page - to `http://sho.rt/urls/ABCD123`. And that will work.

However, if you try to type `ABCD123` in your browser's address bar, nothing will work. There's no domain named ABCD123. You will likely get Google search for that term - https://www.google.com/search?q=ABCD123. This means that you cannot send me `ABCD123` and expect that I'll be redirected to http://www.thelongestdomainnameintheworldandthensomeandthensomemoreandmore.com/. Instead, this code needs to be prefixed with your app's domain name. So, the actual answer is `http://sho.rt/ABCD123`. This new URL now has 21 characters which is much better.

Reply:
This means that our short URLs in the form of `http://sho.rt/ABCD123`. But, can we do better? The goal is to create as short URL as possible. Obviously, we cannot influence the domain name part, all short names are already taken. However, we can decide how long will be that `short` value. Can it be shorter? How many different combinations can we generate?

Reply:
Yes, we decide how long it will be, i think optimal is 5 or 6 characters, with 5 caracters we can have 916.132.832 combinations, its 62*62*62*62*62, 5 times. Why 62? Because we have 26 A-Z characters, 26 a-z characters and 0-9, its 26+26+10=62!

Reply:
62x62x62x62x62*

Reply:
Is it okay if I add "http://" prefix in details blade? Because i can only get localhost:8000 without http when i for example reload or change address and get it back, i lose that header.

Reply:
If you try to generate short url route, it will return the full url, including the `http://` protocol, there's no need to hardcode prefix in your blade files.